README.md
# NiceSnippet-Postman-Codegen

A *hacky?* tool to get Postman to generate custom code without diving into the internals.

## How do I use this?
### Method 1:
1. Get/Open a successfully working postman request.
2. Use Pre-Requisite tab to show code editor
3. Run the code `console.log(pm.request)`.
4. Open up console. Copy output and paste into webapp.
5. Generated code shows up on output box.


### Method 2
1. Get/Open a successfully working postman request.
2. Upload scripts on google appscript.
3. Deploy as Webapp/API.
4. Post Postman request to the API as Json stringified object.
5. Generated code shows up on Response box.


> Written with [StackEdit](https://stackedit.io/).


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project tried to adhere to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.0.0] - 2023-02-26

Major rewrite.

### Added
- New webpage to test input.

### Changed
- Rewrote the codebase from scratch.

## [0.6.4] - 2022-08-05

### Added
- Add check rules function to handle scenerios which Studio cannot handle.
- Check Rule Function - GET request with a body is unsupported

### Changed
- util.js - hasData.body function has been updated to validate data inside object

src/main.js
import { loadMockRequests } from './mockData';
import { hasData, doubleParse } from './utils';

var DEBUG_PRINT = false;
var version = '0.6.4';
var showSystemHeaders = false;

var prefix = 'REQUEST';

var pm_request;
var insidePostman = false;
var UnsupportedFeatureFoundflag = false;
var UnsupportedFeatureFoundLocation;
var sampleCode = '';
var finalProtocol, finalHost, finalPath, finalQuery, finalBody, finalHeader, finalAuth, finalMethod;

function insidePostmanCheck() {
    return typeof pm !== 'undefined';
}

let options = {
    weekday: 'long',
    year: 'numeric',
    month: 'long',
    day: 'numeric',
    timeZoneName: 'short',
    hour: 'numeric',
    minute: 'numeric',
    second: 'numeric',
};
var startTimeStamp = new Date().toLocaleString('en-US', options);

function main() {
    // Check if we are inside postman
    if (insidePostmanCheck()) {
        insidePostman = true;
        loadDataFromPostman();
    } else {
        console.clear(); // does not work inside postman
        pm_request = loadMockRequests();

        console.log('//////////////////// SAMPLE DATA ////////////////////');
        console.log(pm_request);
        console.log('//////////////////// SAMPLE DATA ////////////////////');
        loadDataFromSample(pm_request);
    }
    // Code Execution Workflow
    // displayDataFromMemory()
    buildHeader();
    if (checkRules()) {
        buildCode();
        buildFooter();
        printSampleCode();
    }
}

function loadDataFromSample(request) {
    DEBUG_PRINT && console.log('Running loadDataFromSample');
    // if (request.includes('\\')) {request.replaceAll('\\','\\\\\')}
    request = JSON.parse(request);
    DEBUG_PRINT && console.log({ request });

    // request.url
    // url.protocol - string
    // console.log(request.url.protocol)
    finalProtocol = request.url.protocol;

    // method - string
    finalMethod = request.method;

    // url.host (combine) - array -> string
    // console.log(request.url.host)
    finalHost = '';
    let host = request.url.host;
    host.forEach((h, i) => {
        if (i < host.length - 1) {
            finalHost += h.concat('.');
        } else {
            finalHost += h;
        }
    });
    // DEBUG_PRINT && { console.log({finalHost}) }

    // url.path (combine) - array -> string
    // console.log(request.url.path)
    finalPath = request.url.path;
    // let path = request.url.path
    // path.forEach((p,i) => {
    //     if(i < path.length - 1) { finalPath += p.concat("/") }
    //     else { finalPath += p }
    // });
    // DEBUG_PRINT && { console.log({finalPath})}

    // url.query (combine) - JsonObject -> string
    // console.log(request.url.query)
    if (request.url.query && request.url.query.length > 0) {
        finalQuery = '';
        let query = request.url.query;
        query.forEach((q, i) => {
            finalQuery += q.key;
            finalQuery += '=';
            finalQuery += q.value;
            if (i < query.length - 1) {
                finalQuery += '&';
            }
        });
        finalQuery = finalQuery.split('&');
        // console.log({finalQuery})
    }

    // request.header - array
    // console.log(request.header)
    finalHeader = request.header;
    // console.log({finalHeader})

    // request.auth - JsonObject
    // console.log(request.auth)
    finalAuth = request.auth;
    // console.log(finalAuth)

    // request.body - JsonObject
    // console.log(request.body)
    finalBody = request.body;
    // console.log({finalBody})
}

function loadDataFromPostman() {
    try {
        finalProtocol = pm.request.url.protocol.toString();
    } catch (e) {
        DEBUG_PRINT && console.log('FinalProtocol not found - ', e);
        finalProtocol = '';
    }
    finalHost = pm.request.url.host.join('.');
    finalPath = pm.request.url.path; // array
    finalQuery = pm.request.url.query.toString().split('&'); // array
    finalHeader = doubleParse(pm.request.headers); // object
    finalMethod = pm.request.method; // string
    try {
        finalAuth = pm.request.auth.toJSON(); // json
    } catch (e) {
        DEBUG_PRINT && console.log('FinalAuth not found - ', e);
        finalAuth = '';
    }
    finalBody = pm.request.body;
}

function displayDataFromMemory() {
    DEBUG_PRINT && console.log('//////////////////// displayDataFromMemory ////////////////////');
    DEBUG_PRINT && console.log({ insidePostman });
    DEBUG_PRINT && console.log({ finalProtocol }, typeof finalProtocol, '// string');
    DEBUG_PRINT && console.log({ finalMethod }, typeof finalMethod, '// string');
    DEBUG_PRINT && console.log({ finalHost }, typeof finalHost, '// string');
    DEBUG_PRINT && console.log({ finalPath }, typeof finalPath, '// array?');
    DEBUG_PRINT && console.log({ finalQuery }, typeof finalQuery, '// array?');
    DEBUG_PRINT && console.log({ finalHeader }, typeof finalHeader, '// object?');
    DEBUG_PRINT && console.log({ finalAuth }, typeof finalAuth, '// jsonObject?');
    DEBUG_PRINT && console.log({ finalBody }, typeof finalBody, '// object?');
    DEBUG_PRINT && console.log('//////////////////// displayDataFromMemory ////////////////////');
}

function checkRules() {
    let checkRulesMsg = '';
    let checkRulesStatus = true;
    if (finalMethod === 'GET' && hasData.body(finalBody)) {
        checkRulesMsg = 'Method and Body data combinations are incompatible in studio \n';
        checkRulesMsg += JSON.stringify({ finalMethod }) + '\n';
        checkRulesMsg += JSON.stringify({ finalBody });

        checkRulesStatus = false;
    }

    if (!checkRulesStatus) {
        let msg = '////////////////////////// RULES ERROR ///////////////////////////' + '\n';
        msg += checkRulesMsg + '\n';
        msg += '////////////////////////// RULES ERROR ///////////////////////////';
        console.error(msg);
        throw new Error('Rules Error - Check console for additional details');
    }
    return checkRulesStatus;
}

function buildCode() {
    // sampleCode = ''

    // start
    sampleCode += 'ASSIGN ' + prefix + ' = GetRestProxy()\n';

    // url
    sampleCode += '\n// SECTION: URL\n';
    // -- base
    let urlBase = prefix + '_URL_BASE';
    if (finalProtocol && finalHost) {
        sampleCode += 'ASSIGN ' + urlBase + '   = "' + finalProtocol + '://' + finalHost + '"\n';
    } else if (finalHost) {
        sampleCode += 'ASSIGN ' + urlBase + '   = "' + finalHost + '"\n';
    }
    // -- path
    let urlPath = prefix + '_URL_PATH';
    if (finalPath && finalPath.length > 0) {
        sampleCode += 'ASSIGN ' + urlPath + '   = "';
        finalPath.forEach((pathLine, i) => {
            if (i < finalPath.length - 1) {
                sampleCode += pathLine + '/';
            } else {
                sampleCode += pathLine;
            }
        });
        sampleCode += '"\n';
    }
    // -- method
    let urlMethod = prefix + '_URL_METHOD';
    if (finalProtocol) {
        sampleCode += 'ASSIGN ' + urlMethod + ' = "' + finalMethod + '"\n';
    }
    // -- query
    // TODO: Rewrite this to add appending multiple lines
    // console.log({finalQuery})
    let urlQuery = prefix + '_URL_QUERY';
    if (hasData.query(finalQuery)) {
        //xx
        sampleCode += 'ASSIGN ' + urlQuery + '  = $"';
        finalQuery.forEach((queryLine, i) => {
            if (i < finalQuery.length - 1) {
                // fix single quote
                if (queryLine.indexOf("'") >= 0) {
                    queryLine = queryLine.replaceAll("'", "\\'");
                }
                // TODO: Do I need to fix double quotes and others?
                sampleCode += queryLine + '&';
            } else {
                sampleCode += queryLine;
            }
        });
        sampleCode += '"\n';
    }

    // -- final url combine code
    sampleCode += 'ASSIGN ' + prefix + '_URL        = "{' + urlBase + '}/{' + urlPath + '}';
    if (hasData.query(finalQuery)) {
        sampleCode += '?{' + urlQuery + '}';
    }
    sampleCode += '"\n';

    // header
    // console.log({finalHeader})
    if (hasData.header(finalHeader, showSystemHeaders)) {
        sampleCode += '\n// SECTION: HEADER\n';

        finalHeader.forEach((headerLine, i) => {
            // if ()
            // console.log(h)
            if (headerLine.key.includes('Content-Type')) {
                sampleCode += prefix + '.ContentType = "' + headerLine.value + '"\n';
            } else if (headerLine.hasOwnProperty('system')) {
                if (showSystemHeaders) {
                    sampleCode += '// // Postman Hidden System Headers \n';
                    sampleCode +=
                        '//   ' +
                        prefix +
                        '.AddHeader("' +
                        headerLine.key +
                        '","' +
                        headerLine.value +
                        '")\n';
                }
                // else do not show anything
            } else {
                sampleCode +=
                    prefix + '.AddHeader("' + headerLine.key + '","' + headerLine.value + '")\n';
            }
        });
    }

    // auth
    // console.log({finalAuth})
    if (finalAuth) {
        sampleCode += '\n// SECTION: AUTH\n';

        let username, password;
        if (finalAuth.type === 'basic') {
            finalAuth.basic.forEach((b, i) => {
                if (b.key === 'password') {
                    password = b.value;
                }
                if (b.key === 'username') {
                    username = b.value;
                }
            });
            sampleCode += 'ASSIGN ' + prefix + '_USERNAME = "' + username + '"\n';
            sampleCode += 'ASSIGN ' + prefix + '_PASSWORD = "' + password + '"\n';
            // GetRESTProxy().EncodeBase64("{username}:{password}")
            sampleCode +=
                'ASSIGN ' +
                prefix +
                '_BASICAUTHCREDS = GetRESTProxy().EncodeBase64("{' +
                prefix +
                '_USERNAME' +
                '}';
            sampleCode += ':';
            sampleCode += '{' + prefix + '_PASSWORD' + '}';
            sampleCode += '")\n';
            sampleCode +=
                prefix +
                '.AddHeader("Authorization","Basic {' +
                prefix +
                '_BASICAUTHCREDS}")' +
                '\n';
        }
        if (finalAuth.type === 'bearer') {
            // console.log(finalAuth)
            // finalAuth.bearer
            sampleCode += 'ASSIGN ' + prefix + '_AUTHTOKEN = "' + finalAuth.bearer[0].value + '"\n';
            sampleCode +=
                prefix + '.AddHeader("Authorization","Bearer {' + prefix + '_AUTHTOKEN' + '}")\n';
        } else {
            UnsupportedFeatureFoundflag = true;
            UnsupportedFeatureFoundLocation = 'AUTH';
        }
    }

    // body
    if (hasData.body(finalBody)) {
        DEBUG_PRINT && console.log(finalBody);
        sampleCode += '\n// SECTION: BODY\n';
        if (finalBody.mode === 'raw') {
            let finalbodyraw = JSON.stringify(finalBody.raw);
            finalbodyraw = finalbodyraw.replaceAll('{', '\\{');
            // finalbodyraw = finalbodyraw.replaceAll('\,','\\,')
            sampleCode += 'ASSIGN ' + prefix + '_BODY = ""' + '\n';
            // sampleCode += finalbodyraw
            // remove first and last quote
            finalbodyraw = finalbodyraw.substring(1);
            finalbodyraw = finalbodyraw.substring(0, finalbodyraw.length - 1);
            // console.log({finalbodyraw})
            // console.log(finalbodyraw.split('\\r\\n').forEach((line)=>{console.log({line})}))
            finalbodyraw.split('\\r\\n').forEach((line) => {
                sampleCode += prefix + '_BODY.append($"' + line + '")\n';
            });
        } else {
            UnsupportedFeatureFoundflag = true;
            UnsupportedFeatureFoundLocation = 'BODY';
        }
        sampleCode += '\n';
    }

    // trigger request
    sampleCode += '\n// SECTION: TRIGGER REQUEST\n';

    // MakeRestRequest
    sampleCode += 'ASSIGN RESPONSE = ' + prefix + '.MakeRestRequest(';
    // url
    sampleCode += prefix + '_URL' + ', ';
    // body
    if (hasData.body(finalBody)) {
        sampleCode += prefix + '_BODY';
    } else {
        sampleCode += '""';
    }
    sampleCode += ', ';
    // json or xml
    // TODO: perform checks here
    sampleCode += '0' + ', ';
    // http method
    sampleCode += prefix + '_URL_METHOD';
    sampleCode += ')\n';

    // handle response
    sampleCode += '\n// SECTION: HANDLE RESPONSE\n';
    // Tag Response code and desc
    sampleCode += 'ASSIGN ' + 'RESPONSE' + '_STATUSCODE = ' + prefix + '.StatusCode\n';
    sampleCode += 'ASSIGN ' + 'RESPONSE' + '_STATUSDESC = ' + prefix + '.StatusDescription\n';

    sampleCode += '\n';
    sampleCode +=
        'IF (' + 'RESPONSE' + '_STATUSCODE >= 200 & ' + 'RESPONSE' + '_STATUSCODE < 300) {' + '\n';
    sampleCode += '    // ' + '\n';
    sampleCode += '    // Handle Sucessfull Response' + '\n';
    sampleCode += '    // ' + '\n';
    sampleCode += '}' + '\n';
    sampleCode += 'ELSE {' + '\n';
    sampleCode += '   // ' + '\n';
    sampleCode += '   // Handle UnSucessfull Response' + '\n';
    sampleCode += '   // ' + '\n';
    sampleCode += '   //            (╯°□°）╯︵ ┻━┻)' + '\n';
    sampleCode += '   // ASSIGN ERROR_MESSAGE = "We have a Boo Boo! =( "' + '\n';
    sampleCode += '   // ERROR_MESSAGE.throw()' + '\n';
    sampleCode += '}' + '\n';
}

function printPmRequest(insidePostman) {
    if (insidePostman) {
        console.log('//////////////////  SEND THIS TO MARK!  ////////////////////////////');
        console.log('//--LOCATION: ' + UnsupportedFeatureFoundLocation);
        console.log('//--COPY START------------------------------------------------------');
        console.log(JSON.stringify(pm.request));
        console.log('//--COPY END--------------------------------------------------------');
    } else {
        console.log('//--UNSUPPORTED ENTRY FOUND -----------------------------------');
        console.log('//--LOCATION: ' + UnsupportedFeatureFoundLocation);
    }
}

function printSampleCode() {
    if (!UnsupportedFeatureFoundflag) {
        console.log(sampleCode);
    } else {
        printPmRequest(insidePostman);
    }
}

function buildHeader() {
    let copyStart = '//----------------------------COPY START----------------------------//' + '\n';
    let hrLine = '//////////////////////////////////////////////////////////////////////' + '\n';
    let title =
        "//▂▃▅▇█▓▒░ Mark's Nice Sudio Snippet CodeGen 【ツ】 v" + version + ' ░▒▓█▇▅▃▂' + '\n';
    let brLine = '\n';
    sampleCode += copyStart + hrLine + title + hrLine + brLine;
    if (insidePostman) {
        sampleCode += '// ' + pm.info.requestName + '\n\n';
    }
}

function buildFooter() {
    let copyEnd = '//----------------------------COPY END------------------------------//' + '\n';
    let hrLine = '//////////////////////////////////////////////////////////////////////' + '\n';
    let msg = '// Generated on: ' + startTimeStamp + '\n';
    sampleCode += hrLine + msg + hrLine + copyEnd;
}

// Run Main function
main();


appscript.js
// https://www.labnol.org/code/19871-get-post-requests-google-script

const doPost = (request = {}) => {
    const { parameter, postData: { contents, type } = {} } = request;
    const { source } = parameter;
  
    let result = generateSnippet(contents)
    return ContentService.createTextOutput(result);
  };
  
function testSnippetCodeGen(){
  let input = "{\"description\":{\"content\":\"The HTTP `POST` request method is meant to transfer data to a server \\n(and elicit a response). What data is returned depends on the implementation\\nof the server.\\n\\nA `POST` request can pass parameters to the server using \\\"Query String \\nParameters\\\", as well as the Request Body. For example, in the following request,\\n\\n> POST /hi/there?hand=wave\\n>\\n> <request-body>\\n\\nThe parameter \\\"hand\\\" has the value \\\"wave\\\". The request body can be in multiple\\nformats. These formats are defined by the MIME type of the request. The MIME \\nType can be set using the ``Content-Type`` HTTP header. The most commonly used \\nMIME types are:\\n\\n* `multipart/form-data`\\n* `application/x-www-form-urlencoded`\\n* `application/json`\\n\\nThis endpoint echoes the HTTP headers, request parameters, the contents of\\nthe request body and the complete URI requested.\",\"type\":\"text/plain\"},\"url\":{\"protocol\":\"https\",\"path\":[\"post\"],\"host\":[\"postman-echo\",\"com\"],\"query\":[],\"variable\":[]},\"method\":\"POST\",\"body\":{\"mode\":\"raw\",\"raw\":\"This is expected to be sent back as part of response body.\"}}"
  let result = generateSnippet(input)
  console.log(result)
}

src/mockData.template.js

export function loadMockRequests() {
    // NOTES:
    // 1 - Use console.log(JSON.stringify(pm.request)) to get the pm_request data structure
    // 2 - If there are forward slashes (\), escape it with another slash (\\)
    
    let pm_request
    // simple echo request
    pm_request = '{"url":{"protocol":"https","path":["get"],"host":["postman-echo","com"],"query":[{"key":"foo1","value":"bar1"},{"key":"foo2","value":"bar2"}],"variable":[]},"header":[{"key":"User-Agent","value":"PostmanRuntime/7.29.0","system":true},{"key":"Accept","value":"*/*","system":true},{"key":"Postman-Token","value":"21fe9f4f-a00b-4898-9fb0-e4da8df6a9b8","system":true},{"key":"Host","value":"postman-echo.com","system":true},{"key":"Accept-Encoding","value":"gzip, deflate, br","system":true},{"key":"Connection","value":"keep-alive","system":true}],"method":"GET","body":{}}'

    return pm_request
}

index.js
function writeLog(text = "Hello World") {
    console.log(text)
  }
  
  function generateSnippet(inputText) {
    console.clear()
    // Frontend protection
    if (typeof inputText === 'object') throw "Input is an object so no go!"
  
    // Check if we have valid json
    try {
        inputText = JSON.parse(inputText)
    } catch (error) {
        let msg = "ERROR: generateSnippet - JSON Parse error from input!"
        console.error(msg)
        return msg
        throw msg
    }
    // console.log({inputText})
  
    let sections = Object.keys(inputText)
    let urlCode = headerCode = authCode = methodCode = bodyCode = triggerCode = ""
    sections.forEach((section) => {
        if (section.toLowerCase()==="url") { urlCode = genUrlCode(inputText[section])}
        if (section.toLowerCase()==="header") { headerCode = genHeaderCode(inputText[section])}
        if (section.toLowerCase()==="auth") { authCode = genAuthCode(inputText[section])}
        if (section.toLowerCase()==="method") { methodCode = genMethodCode(inputText[section])}
        if (section.toLowerCase()==="body") { bodyCode = genBodyCode(inputText[section])}
    })
    
    triggerCode = genTriggerCode()
    
    // incase there are no headers
    if (!headerCode) {
        headerCode = genHeaderCode({})
    }
  
    // check for errors
    let sectionVariables = [urlCode, headerCode, authCode, methodCode, bodyCode]
    for (let codeResult in sectionVariables) {
        // console.log(sectionVariables[codeResult])
        if (sectionVariables[codeResult].startsWith('ERROR')) {
            return sectionVariables[codeResult]
        }
    }
    
    let finalCode = '//\n// Url code\n//\n'
    finalCode += urlCode
    finalCode += '\n\n//\n// Header code\n//\n'
    finalCode += headerCode
    if (authCode) {
        finalCode += '\n\n//\n// Auth code\n//\n'
        finalCode += authCode
    }
    finalCode += '\n\n//\n// Method code\n//\n'
    finalCode += methodCode
    finalCode += '\n\n//\n// Body code\n//\n'
    finalCode += bodyCode
    finalCode += '\n\n//\n// Trigger code\n//\n'
    finalCode += triggerCode
    // console.log(finalCode)
    
    // let outputText = JSON.stringify(inputText,null,2)
    return finalCode
  }
  
  function genUrlCode(jsonObj) {
    // console.log(jsonObj)
    let url = getUrl(jsonObj)
    let queryParamsArray = getQueryString(jsonObj)
    // TODO: Handle variable code
    let urlCode = 'ASSIGN RestProxy_Url = "' + url + '"'
    // console.log({queryParamsArray})
    // Check if there is query params
    if (queryParamsArray) {
        
        urlCode += "\n\n// Url Query Params"
        urlCode += '\nASSIGN RestProxy_UrlQueryParams = "?"'
        // add key value of query paramters as a new line
        queryParamsArray.forEach((keyValueString,i) => {
            if (i>0) {
                urlCode += '\nRestProxy_UrlQueryParams.append("&' + keyValueString + '")'
            } else {
                urlCode += '\nRestProxy_UrlQueryParams.append("' + keyValueString + '")'
            }
        })
  
        // Add the query params variable back to the regular URL
        urlCode += '\n\nRestProxy_Url.append(RestProxy_UrlQueryParams)'
    }
    let postmanVars = findInsideCurly(urlCode)
    if (postmanVars.length > 0) {
        urlCode += '\n\n// Replace variables'
        postmanVars.forEach(postmanVar => {
            urlCode += '\nRestProxy_Url = RestProxy_Url.replace("{{' + postmanVar + '}}", "__REPLACE__ME__")'
        })
    }
    urlCode += '\n\n// Debug\nTRACE "RestProxy_Url = {RestProxy_Url}"'
    return urlCode
  }
  
  function getUrl(jsonObj) {
    let protocol = jsonObj["protocol"] || "https"
    let host = jsonObj["host"].join(".")
    if (jsonObj["path"] !== undefined && jsonObj["path"].length > 0) {
        var path = jsonObj["path"].join("/") //optional
    }
  
    let fullurl = protocol
    fullurl += "://"
    fullurl += host
    fullurl += "/"
    if (path) {
        fullurl += path
    }
    // code += "/"  // dont think we need this
  
    // console.log({protocol,path,host,query,code})
    // console.log({fullurl})
    return fullurl
  }
  
  function getQueryString(jsonObj) {
    if (jsonObj["query"] === undefined) {
        return null
    }
    if (jsonObj["query"].length <= 0) {
        return null
    }
    // console.log({queryArrayOfObjects})
    queryArrayOfObjects = jsonObj["query"]
    // console.log({queryArrayOfObjects})
    let queryKvArray = []
    queryArrayOfObjects.forEach((queryKeyValue,i) => {
        let key = queryKeyValue['key']
        let value = queryKeyValue['value']
        // console.log({key,value})
        let keyValueString = key + "=" + value
        // console.log({keyValueString})
        queryKvArray.push(keyValueString)
    })
    // console.log({queryString})
    return queryKvArray
  }
  
  function genHeaderCode(jsonObj){
    // console.log({jsonObj}) // Array of Object
    // discoveryProxy.ContentType = "application/json"
    // skillProxy.AddHeader("Authorization" , "Bearer {bearerToken}")
    let headerCode = ""
    let acceptType = 0 // json
    headerCode += 'ASSIGN RestProxy = GetRESTProxy()'
    if (Object.keys(jsonObj).length === 0 || Object.keys(jsonObj).length === undefined) {
        console.log("genHeaderCode - no header length")
        return headerCode
    }
    // try {
    //     if (jsonObj.length === 0) {
    //         console.log("no length")
    //         return headerCode
    //     }
    // } catch (error) {
    //     console.log({error})
    //     return headerCode
    // }
    jsonObj.forEach((header,i) => {
        if (header.key.toLowerCase() === 'content-type') {
            headerCode += '\nRestProxy.ContentType = "' + header.value + '"'
            if (header.value.includes('xml')) { acceptType = 1 }
        } else {
            headerCode += '\nRestProxy.AddHeader("' + header.key + '", \'' + header.value + '\') '
        }
    })
    // console.log(headerCode)
    headerCode += '\nRestProxy_AcceptType = ' + acceptType + ' // 0 - json, 1 - xml'
  
    return headerCode
  }
  
  function genMethodCode(jsonObj) {
    // console.log({jsonObj})
    // Only allowed - GET, POST, PUT, DELETE
    let allowedVerbs = ['GET','POST','PUT','DELETE']
    if (allowedVerbs.indexOf(jsonObj.toUpperCase()) <= -1) {
        return 'ERROR: genMethodCode - Unsupported verb. Only allow GET,POST,PUT,DELETE\n\n' + JSON.stringify(jsonObj)
        throw 'Unsupported verb. Only allow GET,POST,PUT,DELETE'
    }
    let methodCode = 'ASSIGN RestProxy_Method = "' + jsonObj + '"'
    return methodCode
  }
  
  function genBodyCode(jsonObj) {
    // console.log({jsonObj})
    // console.log(Object.keys(jsonObj).length)
    let bodyCode = 'ASSIGN RestProxy_Body = ""'
    let postmanVars = []
    if (Object.keys(jsonObj).length === 0 || Object.keys(jsonObj).length === undefined) {
        return bodyCode
    }
    if (jsonObj.mode.toLowerCase() === 'raw') {
        // https://stackoverflow.com/questions/21895233/how-to-split-string-with-newline-n-in-node
        let lines = jsonObj.raw.split(/\r?\n/)
        lines.forEach((line,i) => {
            // check for variables
            if (findInsideCurly(line).length > 0) {
                postmanVars.push(findInsideCurly(line))
            }
  
            // Add lines to request body
            let escapeLine = escapeStrings(line)
            // console.log({escapeLine})
            bodyCode += '\nRestProxy_Body.append($' + JSON.stringify(escapeLine + insertNewLine()) + ')'
            // bodyCode += '\nRestProxy_Body.append($"{CHAR(10)}")'
        })
  
        // check for xml or json - to set restproxy accept type
        // 0 for json, 1 for xml
        // TODO: Change this to use contentType
        let lang = ""
        try {
            lang = jsonObj.options.raw.language
        } catch (error) {
            lang = 'json'
        }
        // console.log({lang})
        if (lang.toLowerCase().includes("xml")) {
            bodyCode += '\nRestProxy_AcceptType = 1 // 0 - json, 1 - xml'
        } else {
            bodyCode += '\nRestProxy_AcceptType = 0 // 0 - json, 1 - xml'
        }
  
        // check for variables
        postmanVars = postmanVars.flat(1)
        // console.log({postmanVars})
        if (postmanVars.length > 0) {
            bodyCode += '\n\n// Replace variables'
            postmanVars.forEach(postmanVar => {
                bodyCode += '\nRestProxy_Body = RestProxy_Body.replace("{{' + postmanVar + '}}", "__REPLACE__ME__")'
            })
        }
  
        // TRACE "RestProxy_Body = {RestProxy_Body}"
        bodyCode += '\n\n// Debug\nTRACE "RestProxy_Body = {RestProxy_Body}"'
    } else {
        return 'ERROR: genBodyCode - Unhandled Body mode!\n\n' + JSON.stringify(jsonObj)
        throw 'Unhandled Body mode!'
    }
    // console.log(bodyCode)
    return bodyCode
  }
  
  function insertNewLine() {
    // https://www.cs.cmu.edu/~pattis/15-1XX/common/handouts/ascii.html
    let enable = true
    if (enable) {
        return "{CHAR(10)}"
    } else {
        return ''
    }
  }
  
  function findInsideCurly(line) {
    // Check for postman vars
    // https://stackoverflow.com/questions/413071/regex-to-get-string-between-curly-braces
    // [^{\}]+(?=})
    // let findInsideCurly = /[^{\}]+(?=})/g // single curly
    let findInsideCurly = /[^{{\}]+(?=}})/g // double curly
    let result = []
    let match = []
    let matchLength = ""
    if (findInsideCurly.test(line)) {
        match = line.match(findInsideCurly)
        matchLength = match.length
        // console.log({match,matchLength})
    }
    // if (matchLength > 0) {
    //     console.log({match})
    // }
    return match
  }
  
  function escapeStrings(line) {
    let result = JSON.stringify(line)
    result = line.substring(1)
    result = line.substring(-1)
    // https://help.nice-incontact.com/content/studio/scripting/reference/stringhandling.htm?Highlight=escape
    // result =    line.replace(/"/g, '\\"')    //  "  -->  \"
    //                     .replace(/'/g, "\\'")    //  '  -->  \'
    //                     .replace(/\"/g, '\\"')    //  '  -->  \'
    //                     // .replace(/\\/g, '\\\\')  //  \\ --> \\\\
    //                     // .replace(/\n/g, '\\n')   // new line
    //                     // .replace(/\r/g, '\\r');  // line feed
  
    // skip {{var}} as postman variables
    if (!(line.includes("{{") && line.includes("}}"))) {
        result = line.replace(/{/g, '\\{')    //  {  -->  \{
    }
  
    // return result
    return line
  }
  
  function genTriggerCode() {
    let triggerCode = 'ASSIGN RestProxy_Response = RestProxy.MakeRestRequest(RestProxy_Url, "", RestProxy_AcceptType, RestProxy_Method)'
    triggerCode += '\nRestProxy_Response_StatusCode = RestProxy.StatusCode'
    triggerCode += '\nRestProxy_Response_StatusDesc = RestProxy.StatusDescription'
    triggerCode += '\n\n// Debug'
    triggerCode += '\nTRACE "RestProxy_Response_StatusCode = {RestProxy_Response_StatusCode}"'
    triggerCode += '\nTRACE "RestProxy_Response_StatusDesc = {RestProxy_Response_StatusDesc}"'
    triggerCode += '\nTRACE "RestProxy_Response = {RestProxy_Response1.asjson()}"'
    triggerCode += '\n\nIF (RestProxy_Response_StatusCode >= 200 & RestProxy_Response_StatusCode < 300)'
    triggerCode += '\n{'
    triggerCode += '\n    // '
    triggerCode += '\n    // Handle Sucessfull Response'
    triggerCode += '\n    // '
    triggerCode += '\n}'
    triggerCode += '\nELSE {'
    triggerCode += '\n   // '
    triggerCode += '\n   // Handle Unsucessfull Response'
    triggerCode += '\n   // '
    triggerCode += '\n   //            (╯°□°）╯︵ ┻━┻)'
    triggerCode += '\n   // ASSIGN ERROR_MESSAGE = "We have a Boo Boo! =( "'
    triggerCode += '\n   // ERROR_MESSAGE.throw()'
    triggerCode += '\n}'
  
    return triggerCode
  }
  
  
  function genAuthCode(jsonObj) {
    let authCode = ""
    // console.log({jsonObj})
    if (jsonObj.type.toLowerCase() === 'basic') {
        let username = password = ""
        jsonObj.basic.forEach(item => {
            if (item.key.toLowerCase() === 'username') {username = item.value}
            if (item.key.toLowerCase() === 'password') {password = item.value}
        })
        // console.log({username,password})
        let basicAuth = 'Basic ' + btoa(username + ':' + password)
        console.log({basicAuth})
        authCode += 'RestProxy.AddHeader("Authorization","' + basicAuth + '")'
        return authCode
    } else {
        return 'ERROR: genAuthCode - Unsupported Auth mode\n\n' + JSON.stringify(jsonObj)
        throw 'Unsupported Auth mode'
    }
  }